### PR TITLE
New Retry Strategy With Exponential Backoff

### PR DESCRIPTION
**Pull request by liliyang**
**Goals (and why)**:
New (configurable) transaction retry strategy with randomized exponential backoff. 

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    @Before
+    public void before() {
+        when(random.nextLong(anyLong())).thenAnswer(inv -> (long) inv.getArgument(0) - 1);
+        when(shouldStopRetrying.test(anyInt())).thenAnswer(inv -> ((int) inv.getArgument(0)) > 1);
+        legacy = TransactionRetryStrategy.createLegacy(blockStrategy);
+        exponential = TransactionRetryStrategy.createExponential(blockStrategy, random);
+
+    }
+
+    private String run() throws Exception {
```
Comment (felixdesouza) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: `run` -> `runExponential`

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        return exponential.runWithRetry(shouldStopRetrying, task);
+    }
+
+    @Test
+    public void retriesIfFailsWithRetriableException() throws Exception {
+        when(task.run()).thenThrow(new TransactionFailedRetriableException("")).thenReturn("success");
```
Comment (felixdesouza) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: is there an easy of checking whether or not we retried at least once (or twice depending on how you define retries)

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        assertThatExceptionOfType(TransactionFailedRetriableException.class)
+                .isThrownBy(this::run)
+                .withMessage("Failing after 2 tries.")
+                .withCause(second);
+    }
+
+    @Test
+    public void doesNotRetryOnNonRetriableTransactionFailedException() throws Exception {
+        TransactionFailedNonRetriableException failure = new TransactionFailedNonRetriableException("");
+        when(task.run()).thenThrow(failure);
```
Comment (felixdesouza) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: can we add a successful response after this? shows that we don't bother retrying after the fact

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    @Test
+    public void rethrowsErrors() throws Exception {
+        AssertionError error = new AssertionError();
+        when(task.run()).thenThrow(error);
+        assertThatThrownBy(this::run).isEqualTo(error);
+    }
+
+    @Test
+    public void doesNotBackOffIfLegacy() throws Exception {
+        when(shouldStopRetrying.test(anyInt())).thenAnswer(inv -> ((int) inv.getArgument(0)) > 50);
```
Comment (felixdesouza) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: not really sure what's going on here, can we move it to a method, so it clears up the individual test, we have it in a few other places in the test

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        assertThatExceptionOfType(TransactionFailedRetriableException.class)
+                .isThrownBy(() -> legacy.runWithRetry(shouldStopRetrying, task));
+        assertThat(blockStrategy.totalBlockedTime).isZero();
+    }
+
+    @Test
+    public void backsOffExponentially() throws Exception {
+        when(shouldStopRetrying.test(anyInt())).thenAnswer(inv -> ((int) inv.getArgument(0)) > 11);
+        when(task.run()).thenAnswer(inv -> {
+            System.out.println(blockStrategy.totalBlockedTime);
```
Comment (felixdesouza) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: remove `println`

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java (hunk 1 above)
Comment (felixdesouza) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: have a corresponding `runLegacy`, or it that doesn't make sense probably just get rid of this method

File: atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java (hunk 3 above)
Comment (j-baker) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: think you'll find that Mockito disallows that?

Comment (j-baker) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java]: it makes sense to add a verify

**Pull request merged**